storage/client/internal/curl_request.cc

```diff
@@ -0,0 +1,103 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/client/internal/curl_request.h"
+#include "storage/client/internal/curl_wrappers.h"
+#include "google/cloud/internal/build_info.h"
+#include "google/cloud/internal/throw_delegate.h"
+#include <iostream>
+#include <sstream>
+
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+
+CurlRequest::CurlRequest(std::string base_url)
+    : url_(std::move(base_url)),
+      query_parameter_separator_("?"),
+      curl_(curl_easy_init()),
+      headers_(nullptr) {
+  if (curl_ == nullptr) {
+    google::cloud::internal::RaiseRuntimeError("Cannot initialize CURL handle");
+  }
+}
+
+CurlRequest::~CurlRequest() {
+  curl_slist_free_all(headers_);
+  curl_easy_cleanup(curl_);
+}
+
+void CurlRequest::AddHeader(std::string const& key, std::string const& value) {
+  AddHeader(key + ": " + value);
+}
+
+void CurlRequest::AddHeader(std::string const& header) {
+  headers_ = curl_slist_append(headers_, header.c_str());
+}
+
+void CurlRequest::AddQueryParameter(std::string const& key,
+                                    std::string const& value) {
+  std::string parameter = query_parameter_separator_;
+  parameter += MakeEscapedString(key).get();
+  parameter += "=";
+  parameter += MakeEscapedString(value).get();
+  query_parameter_separator_ = "&";
+  url_.append(parameter);
+}
+
+/**
+ * Make a request with the given payload.
+ *
+ * @param payload The contents of the request.
+ * @return The response payload as a string.
+ */
+std::string CurlRequest::MakeRequest(std::string const& payload) {
+  // Pre-compute and cache the user agent string:
+  static std::string const user_agent = [] {
```

Comment:
Probably not super important for the initial version of this class, but users often want to use a custom user agent (IIRC, we tack their custom string on the end of our standardized user agent... but I'd have to check with the client libraries folks).  Might be useful to allow setting a custom user agent segment elsewhere and include it in this string - we can make a TODO for that and shelve it for now though.

Reply:
Ack, created #511 to track this, and added a placeholder into the design document.